src/dft/src/stitch/ScanStitch.cpp

```diff
@@ -0,0 +1,182 @@
+constexpr std::string_view kScanOutNamePattern = "scan_out_{}";
+
+}  // namespace
+
+namespace dft {
+
+ScanStitch::ScanStitch(odb::dbDatabase* db) : db_(db)
+{
+  odb::dbChip* chip = db_->getChip();
+  top_block_ = chip->getBlock();
+}
+
+void ScanStitch::Stitch(
+    const std::vector<std::unique_ptr<ScanChain>>& scan_chains)
+{
+  // TODO: For now, we only use one scan enable for all the chains. We may
+  // support in the future multiple test modes
+  if (scan_chains.empty()) {
+    return;
+  }
+
+  odb::dbChip* chip = db_->getChip();
+  odb::dbBlock* block = chip->getBlock();
+  FindOrCreateScanEnable(block);
+
+  for (const std::unique_ptr<ScanChain>& scan_chain : scan_chains) {
+    Stitch(block, *scan_chain);
+  }
+}
+
+void ScanStitch::Stitch(odb::dbBlock* block, const ScanChain& scan_chain)
+{
+  // Let's create the scan in and scan out of the chain
+  // TODO: Suport user defined scan signals
+
+  ScanDriver scan_in_port = FindOrCreateScanIn(block);
+
+  // We need fast pop for front and back
+  std::deque<std::shared_ptr<ScanCell>> scan_cells;
+
+  const std::vector<std::shared_ptr<ScanCell>>& falling_edge
+      = scan_chain.getFallingEdgeScanCells();
+  const std::vector<std::shared_ptr<ScanCell>>& rising_edge
+      = scan_chain.getRisingEdgeScanCells();
+
+  // Falling edge first
+  std::copy(
+      falling_edge.begin(), falling_edge.end(), std::back_inserter(scan_cells));
+  std::copy(
+      rising_edge.begin(), rising_edge.end(), std::back_inserter(scan_cells));
+
+  // All the cells in the scan chain are controlled by the same scan enable
+  for (const auto& scan_cell : scan_cells) {
+    scan_cell->connectScanEnable(*scan_enable_);
+  }
+
+  // Lets get the first and last cell
+  const std::shared_ptr<ScanCell>& first_scan_cell = *scan_cells.begin();
+  const std::shared_ptr<ScanCell>& last_scan_cell = *(scan_cells.end() - 1);
+
+  if (!scan_cells.empty()) {
+    scan_cells.pop_front();
+  }
+
+  if (!scan_cells.empty()) {
+    scan_cells.pop_back();
+  }
+
+  for (auto current = scan_cells.begin(); current != scan_cells.end();
+       ++current) {
+    auto next = current + 1;
+    // Connects current cell scan out to next cell scan in
+    (*next)->connectScanIn((*current)->getScanOut());
+  }
+
+  // Let's connect the first cell
+  first_scan_cell->connectScanEnable(*scan_enable_);
+  first_scan_cell->connectScanIn(scan_in_port);
+
+  if (!scan_cells.empty()) {
+    (*scan_cells.begin())->connectScanIn(first_scan_cell->getScanOut());
+  } else {
+    // If last_scan_cell == first_scan_cell, then scan in was already connected
+    if (last_scan_cell != first_scan_cell) {
+      last_scan_cell->connectScanIn(first_scan_cell->getScanOut());
+    }
+  }
+
+  // Let's connect the last cell
+  ScanLoad scan_out_port
+      = FindOrCreateScanOut(block, last_scan_cell->getScanOut());
+  last_scan_cell->connectScanOut(scan_out_port);
+}
+
+void ScanStitch::FindOrCreateScanEnable(odb::dbBlock* block)
```

Comment:
Why doesn't this return anything?

Reply:
It was setting scan_enable_ as a member of the ScanStitch object, but now I am returning a ScanDriver and passing it to the other methods.